packages/@glimmer/integration-tests/test/owner-test.ts

```diff
@@ -81,6 +169,73 @@ class OwnerTest extends RenderTest {
 
     assert.verifySteps(['foo-bar owner called', 'foo-baz owner called'], 'owners used correctly');
   }
+
+  @test
+  'owner can be changed by a component with the hasSubOwner capability'(assert: Assert) {
+    let owner1 = { name: 'owner1' };
+    let owner2 = { name: 'owner2' };
+
+    const CheckOwner2 = defineCheckOwnerComponent(owner1, assert);
+    const CheckOwner1 = defineCheckOwnerComponent(owner1, assert);
+
+    const Mount2 = defineMountComponent(owner2, { CheckOwner2 }, `<CheckOwner2/>{{yield}}`);
+    const Mount1 = defineMountComponent(
+      owner1,
+      { CheckOwner1, Mount2 },
+      `<CheckOwner1/><Mount2><CheckOwner1/></Mount2>`
+    );
+
+    this.renderComponent(Mount1);
+  }
+
+  @test
+  'owner is preserved in curried closure components'(assert: Assert) {
+    let owner1 = { name: 'owner1' };
+    let owner2 = { name: 'owner2' };
+
+    const CheckOwner2 = defineCheckOwnerComponent(owner1, assert);
+    const CheckOwner1 = defineCheckOwnerComponent(owner1, assert);
+
+    const Mount2 = defineMountComponent(owner2, { CheckOwner2 }, `<CheckOwner2/><@CheckOwner1/>`);
+    const Mount1 = defineMountComponent(
+      owner1,
+      { CheckOwner1, Mount2 },
+      `<CheckOwner1/><Mount2 @CheckOwner1={{component CheckOwner1}}/>`
+    );
+
+    this.renderComponent(Mount1);
+  }
+
+  // TODO: This behavior could be confusing the users, but currently we don't know of a way
+  // to ensure we are using the component with the correct owner if it was not curried.
+  // We should continue exploring options here.
+  @test
+  'owner is preserved in non-curried component definitions that are passed around'(assert: Assert) {
+    let owner1 = { name: 'owner1' };
+    let owner2 = { name: 'owner2' };
+
+    const CheckOwner2 = defineCheckOwnerComponent(owner1, assert);
+    const CheckOwner1 = defineCheckOwnerComponent(owner1, assert);
+
+    const Mount2 = defineMountComponent(owner2, { CheckOwner2 }, `<CheckOwner2/><@CheckOwner2/>`);
+    const Mount1 = defineMountComponent(
+      owner1,
+      { CheckOwner1, CheckOwner2, Mount2 },
+      `<CheckOwner1/><Mount2 @CheckOwner2={{CheckOwner2}}/>`
+    );
+
+    this.renderComponent(Mount1);
+  }
```

Comment:
For context, we (@pzuraq + @krisselden + myself) discussed this in detail. The rough conclusion is that we don't really like the thing this test is asserting (it is very strange to end up with an owner from a different invocation than where you were originally associated), but it is quite difficult to ensure that we **always** have the right owner (it isn't even possible to absorb all variants of things that could be done in JavaScript land).

At the moment, we intend to land this update over in Ember and add a debug only AST transform that wraps any values passed to `{{mount 'foo' model=` (because as far as we can tell, that is the only way to pass things "across owner / realm") and asserts if you pass something that is a component/helper/modifier that isn't wrapped in `(component`/`(helper`/`(modifier` helpers.